### PR TITLE
[actions] fix "sh" broken when output is in encoding other than UTF-8

### LINKED ISSUE #15153: Running xcrun extractLocStrings as part of Fastlane causes crash for duplicated high-codepoint unicode localisable strings
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
We run `xcrun extractLocStrings` as part of our fastlane workflow and we have a project where some of the localisable strings contain high-codepoint unicode characters such as "µm" where the `µ` is the high-codepoint character.  This string is localisable in multiple parts of the codebase and by default `xcrun extractLocStrings` outputs standard output warning you that you have used multiple comments for those strings.

When fastlane outputs to screen the first instance of one of the warnings where the localisable string contains a high-codepoint character, it crashes with the stacktrace shown further below.

I was able to supress the output of the warnings from `xcrun extractLocStrings` by using the flag `-q` but this is not the default and fastlane should be able to handle the output generated by any tool.

##### Command executed

`bin/fastlane test`

##### Complete output when running fastlane, including the stack trace and command used

```bash
/Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane_core/lib/fastlane_core/ui/implementations/shell.rb:71:in `split': [!] invalid byte sequence in UTF-8 (ArgumentError)
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane_core/lib/fastlane_core/ui/implementations/shell.rb:71:in `command_output'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane_core/lib/fastlane_core/ui/ui.rb:17:in `method_missing'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/helper/sh_helper.rb:57:in `block (2 levels) in sh_control_output'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/helper/sh_helper.rb:56:in `each'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/helper/sh_helper.rb:56:in `block in sh_control_output'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/open3.rb:205:in `popen_run'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/open3.rb:190:in `popen2e'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/helper/sh_helper.rb:54:in `sh_control_output'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/helper/sh_helper.rb:16:in `sh_no_action'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/fast_file.rb:213:in `block in sh'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/actions/actions_helper.rb:50:in `execute_action'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/fast_file.rb:212:in `sh'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/fast_file.rb:206:in `sh'
	from Fastfile:145:in `block (2 levels) in parsing_binding'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/lane.rb:33:in `call'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/runner.rb:204:in `try_switch_to_lane'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/runner.rb:146:in `trigger_action_by_name'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/fast_file.rb:154:in `method_missing'
	from Fastfile:176:in `block (2 levels) in parsing_binding'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/lane.rb:33:in `call'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/runner.rb:49:in `block in execute'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/runner.rb:45:in `chdir'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/runner.rb:45:in `execute'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/lane_manager.rb:56:in `cruise_lane'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/commands_generator.rb:108:in `block (2 levels) in run'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/commands_generator.rb:333:in `run'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/commands_generator.rb:41:in `start'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/fastlane/lib/fastlane/cli_tools_distributor.rb:119:in `take_off'
	from /Users/my_user/buildAgent/work/1234567890abcdef/_vendor/bundle/ruby/2.3.0/gems/fastlane-2.127.0/bin/fastlane:23:in `<top (required)>'
	from /Users/my_user/buildAgent/work/1234567890abcdef/bin/fastlane:29:in `load'
	from /Users/my_user/buildAgent/work/1234567890abcdef/bin/fastlane:29:in `<main>'
```

### Environment

<details>
  <pre><details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                                          |
| --------------------------- | -------------------------------------------------------------- |
| OS                          | 10.14.6                                                        |
| Ruby                        | 2.3.7                                                          |
| Bundler?                    | true                                                           |
| Git                         | git version 2.20.1 (Apple Git-117)                             |
| Installation Source         | bin/fastlane                                                   |
| Host                        | Mac OS X 10.14.6 (18G87)                                       |
| Ruby Lib Dir                | /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib |
| OpenSSL Version             | LibreSSL 2.6.5                                                 |
| Is contained                | false                                                          |
| Is homebrew                 | false                                                          |
| Is installed via Fabric.app | false                                                          |
| Xcode Path                  | /Applications/Xcode1030.app/Contents/Developer/                |
| Xcode Version               | 10.3                                                           |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_GB.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.128.1 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

| Plugin                    | Version | Update-Status |
| ------------------------- | ------- | ------------- |
| fastlane-plugin-appcenter | 1.2.1   | ✅ Up-To-Date  |

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| bundler                   | 1.16.6       |
| io-console                | 0.4.5        |
| rake                      | 12.3.2       |
| CFPropertyList            | 3.0.0        |
| concurrent-ruby           | 1.1.5        |
| i18n                      | 0.9.5        |
| minitest                  | 5.11.3       |
| thread_safe               | 0.3.6        |
| tzinfo                    | 1.2.5        |
| activesupport             | 4.2.11.1     |
| public_suffix             | 2.0.5        |
| addressable               | 2.6.0        |
| eventmachine              | 1.2.7        |
| websocket-extensions      | 0.1.4        |
| websocket-driver          | 0.7.1        |
| faye-websocket            | 0.10.9       |
| childprocess              | 1.0.1        |
| rubyzip                   | 1.2.3        |
| selenium-webdriver        | 3.142.3      |
| appium_lib_core           | 3.2.0        |
| mini_portile2             | 2.4.0        |
| nokogiri                  | 1.10.3       |
| tomlrb                    | 1.2.8        |
| appium_lib                | 10.3.1       |
| atomos                    | 0.1.3        |
| babosa                    | 1.0.2        |
| claide                    | 1.0.3        |
| colored                   | 1.2          |
| colored2                  | 3.1.2        |
| colorize                  | 0.8.1        |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| declarative               | 0.0.10       |
| declarative-option        | 0.1.0        |
| diff-lcs                  | 1.3          |
| digest-crc                | 0.4.1        |
| unf_ext                   | 0.0.7.6      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20190701 |
| dotenv                    | 2.7.5        |
| emoji_regex               | 1.0.1        |
| excon                     | 0.66.0       |
| multipart-post            | 2.0.0        |
| faraday                   | 0.15.4       |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday_middleware        | 0.13.1       |
| fastimage                 | 2.1.5        |
| gh_inspector              | 1.1.3        |
| jwt                       | 2.1.0        |
| memoist                   | 0.16.0       |
| multi_json                | 1.13.1       |
| os                        | 1.0.1        |
| signet                    | 0.11.0       |
| googleauth                | 0.6.7        |
| httpclient                | 2.8.3        |
| mime-types-data           | 3.2019.0331  |
| mime-types                | 3.2.2        |
| uber                      | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| google-api-client         | 0.23.9       |
| google-cloud-env          | 1.2.0        |
| google-cloud-core         | 1.3.0        |
| google-cloud-storage      | 1.16.0       |
| json                      | 2.2.0        |
| mini_magick               | 4.9.5        |
| multi_xml                 | 0.6.0        |
| plist                     | 3.5.0        |
| security                  | 0.1.3        |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.5        |
| slack-notifier            | 2.3.2        |
| terminal-notifier         | 2.0.0        |
| unicode-display_width     | 1.6.0        |
| terminal-table            | 1.8.0        |
| tty-screen                | 0.7.0        |
| tty-cursor                | 0.7.0        |
| tty-spinner               | 0.9.1        |
| word_wrap                 | 1.0.0        |
| nanaimo                   | 0.2.6        |
| xcodeproj                 | 1.12.0       |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| xcpretty-travis-formatter | 1.0.0        |
| fastlane-plugin-appcenter | 1.2.1        |
| handlebars-source         | 4.0.14       |
| libv8                     | 3.16.14.19   |
| ref                       | 2.0.0        |
| therubyracer              | 0.12.3       |
| handlebars                | 0.8.0        |
| netrc                     | 0.11.0       |
| rest-client               | 2.0.2        |
| rspec-support             | 3.8.2        |
| rspec-core                | 3.8.2        |
| rspec-expectations        | 3.8.4        |
| rspec-mocks               | 3.8.1        |
| rspec                     | 3.8.0        |
| rubyXL                    | 3.4.5        |
| ruby_rncryptor            | 3.0.2        |
</details>

*generated on:* **2019-08-07**
</details></pre>
</details>

### PR DESCRIPTION
**Pull request by sdjmchattie**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

Deals with logging of standard output from executed commands where the output is in encodings other than UTF-8.  Currently only UTF-16 and ISO-8859-1 are being handled and converted back to UTF-8, but these are the most common encodings.  Without this, the built in Ruby split command fails and raises an error.  This at least prevents most of these errors and if it makes a mistake with the encoding conversion then the worst that happens is that some less common characters are displayed incorrectly in the output.

Resolves #15153

### Description

The change is fairly self explanatory, but basically we go through the possible encodings and force that encoding onto the string.  If Ruby then says the string has a valid encoding, we convert from that to UTF-8 for the split command to work.  The provided unit test checks that converting a string from UTF-16 and from ISO-8859-1 do not raise any errors now.  When the fix is removed, this test fails.

### CODE REVIEW

**Comment by sdjmchattie** [comment]
@googlebot I signed it!

**Review by chilch — APPROVED** [review]

**Review by chilch — APPROVED** [review]
Looks good to me, thanks for fix!

**Review by ainame — COMMENTED** [review]
It looks like your test case doesn't look right. The precondition of the scenario is not correct. (I've read through all #15153) What actually happens with the existing code for your issue's scenario is that the `String` object - `message` passed to `FastlaneCore::Shell#command_output` uses UTF-8 as its encoding while the real output has some UTF-16 characters, which is **invalid**. However, your test case prepared `String` objects encoded to UTF-16 or ISO-8859-1, which are both just **valid**.

For example, this runs successfully.

```ruby
p  "string_in_utf-16_with_µ\n".encode('UTF-16').split #=> ["\uFEFFstring_in_utf-16_with_\u00B5"]
```

I would suggest that you use `String#scrub` which eliminates/replaces invalid characters. This is super simple and does work for your scenario, I suppose. Most importantly this can cover all the possible encoding.

```ruby
actual = (message.scrub.split("\r").last || "") # as clearing the line will remove the `>` and the time stamp
```

A better test case being close to the issue's scenario would be like below.

```ruby
# Hard coded UTF-16 string but this String object is based on UTF-8
expect { shell.command_output("\xB5") }.not_to raise_error
```

What do you think?

**Comment by sdjmchattie** [comment]
@ainame when I remove my fix from the code, the test fails. When I run the text file from #15153  through Fastlane (I.e. not in the unit test) it raises the error before my fix but works correctly after my fix. If you use `scrub` you would throw away good characters from the encoding used if you’re scrubbing while the string thinks it’s in UTF-8 encoding. With my fix, the command output contains all the correct characters in the output of the process. I think it’s fair to try to parse the string as UTF-16 if common tools like genstrings are likely to be parsed via Fastlane. It’s preferable to throwing characters away unnecessarily.

**Comment by ainame** [comment]
@sdjmchattie 

> I think it’s fair to try to parse the string as UTF-16 if common tools like genstrings are likely to be parsed via Fastlane. 

I see your point. I missed the output you got were warnings from `genstrings`. Can you please share which encodings are necessary by which tools in your case? It's good to document that in code.

File: fastlane_core/lib/fastlane_core/ui/implementations/shell.rb [hunk 1]
```diff
@@ -150,6 +150,19 @@ def password(message)
...
+
+      # Try some common encodings and return the first with valid codings
+      ['UTF-8', 'UTF-16', 'ISO-8859-1'].each do |dest_encoding|
+        message.force_encoding(dest_encoding)
+        return message.encode('UTF-8', dest_encoding) if message.valid_encoding?
+      end
+
+      message.force_encoding(orig_encoding)
+      return message
+    end
```
Comment (ainame) [inline fastlane_core/lib/fastlane_core/ui/implementations/shell.rb:164]: And also as a piece of Ruby source code, this is not clear to me. I might want to improve this part.

This makes it a bit more Ruby-ish  code with using constant values and Object#dup not to mutate given parameter. Please document the use cases of encodings in the comment 🙇 

```suggestion
    def encode_as_utf_8_if_possible(message)
      return message if message.valid_encoding?
      
      # Try following common encodings and return the first with valid codings
      # 
      # * UTF-16 - used by genstrings
      # * ISO8859-1 - used by blah blah 
      test_string = message.dup
      [Encoding::UTF_16, Encoding::ISO8859_1].each do |dest_encoding|
        return message.encode(Encoding::UTF_8, dest_encoding) if test_string.force_encoding(dest_encoding).valid_encoding?
      end

      # return the original message if no valid encoding is found
      message
    end
```

**Review by ainame — CHANGES_REQUESTED** [review]
Suggested some improvements in code.

**Comment by sdjmchattie** [comment]
Thanks @ainame I was a bit confused by your suggestions at first, but I've puzzled it out.  You are right that the string I generated in the unit test was wrong.  It was valid, as you suggested.  The problem that this is supposed to be resolving is that all standard output from other sources is assumed to be UTF-8 when it sometimes is not.  I've modified the unit test now to create a UTF-16 string (that is valid) and then invalidly force it to UTF-8 so that Ruby now thinks it's UTF-8.  This is the exact scenario that genstrings causes when we had the µ character in one of our strings.  I accepted your change to the make_utf8 method but also tweaked it to remove the ISO-8859-1 which I only included because it is a common encoding.  I don't know of a tool that outputs it specifically, but if you don't want it it wasn't causing a problem for me.

Hopefully the changes I've made will give you the "ahha" moment and that you can also now see that `p "hello\n".encode('UTF-16').force_encoding('UTF-8').split` raises an error.  The error I was seeing for the past 3 years with genstrings output.

**Comment by ainame** [comment]
Can you please fix CI failures?

(Note that on the latest master branch I made a bug that CI checks pass when failing actually😅 but it will be resolved by this soon https://github.com/fastlane/fastlane/pull/18398 )

**Comment by twocountriestrus** [comment]
Gracias

**Comment by rogerluan** [comment]
Just adding my 2 cents:

I think this is very key in the issue:

>I realized that `NSLocalizedString("µ", comment: "")` reproduces a perfectly fine output, whereas if you add a letter `m` in front of it, it glitches, such as: `NSLocalizedString("µm", comment: "")`. Hm, interesting.

We should probably test that specific scenario to really make sure everything's 👌 😄

File: fastlane_core/lib/fastlane_core/ui/implementations/shell.rb [hunk 2]
```diff
@@ -150,6 +150,21 @@ def password(message)
...
     private
 
+    def encode_as_utf_8_if_possible(message)
+      return message if message.valid_encoding?
+
+      # Try following common encodings and return the first with valid codings
+      #
+      # * UTF-16 - output by genstrings
+      test_string = message.dup
+      [Encoding::UTF_16].each do |dest_encoding|
```
Comment (minuscorp) [inline fastlane_core/lib/fastlane_core/ui/implementations/shell.rb:160]: Is this needed when there's only one encoding to be tested on?

Comment (minuscorp) [reply fastlane_core/lib/fastlane_core/ui/implementations/shell.rb:160]: Yeah, that's why I asked of keeping the array now that we only have one left, not that it'll cost much computationally talking😂

**Comment by sdjmchattie** [comment]
I don't know how to fix the CI issues.  Some of these seem to change every time I merge master into this branch.
The previous run only failed on Ruby 2.4 but I can't reproduce the issue locally.
I installed Ruby 2.4.6p354 via `rbenv` so it's the exact same version as on the CI server and the test passes on my machine.

**Comment by minuscorp** [comment]
Ruby 2.4 raises the next backtracked:
```
Failure/Error:
  expect do
    @shell.command_output("utf_16_string_tagged_as_utf_8\n".encode(Encoding::UTF_16).force_encoding(Encoding::UTF_8))
    @shell.command_output("µm\n".encode(Encoding::UTF_16).force_encoding(Encoding::UTF_8))
  end.not_to raise_error

  expected no Exception, got #<ArgumentError: invalid byte sequence in UTF-8>
```

**Comment by sdjmchattie** [comment]
Sorry @minuscorp maybe I've missed the information you're giving me.  Here's what happens when I run this on my machine with the same version of Ruby:

image

I don't know how to fix it if I can't see why it's broken.

**Comment by minuscorp** [comment]
Maybe try to run the whole suite with a random order (see the command in the circleci) to see if it gets raised 🧐

**Comment by ainame** [comment]
I may look into it later 🙂

**Comment by joshdholtz** [comment]
@googlebot I consent.

**Comment by joshdholtz** [comment]
@sdjmchattie @ainame I fixed the failing Ruby 2.4 test by replacing any invalid encoding with empty string 🤷‍♂️ Seemed like this was _only_ an issue in Ruby 2.4 for some unknown reason. Hope this is an okay fix 😇

**Comment by sdjmchattie** [comment]
@googlebot I consent

**Comment by sdjmchattie** [comment]
Thanks @joshdholtz I appreciate that. So weird that it had a problem on the CI server because locally I couldn't reproduce the issue.

File: fastlane_core/lib/fastlane_core/ui/implementations/shell.rb [hunk 3]
```diff
@@ -150,6 +150,17 @@ def password(message)
...
 
+    def encode_as_utf_8_if_possible(message)
+      return message if message.valid_encoding?
+
+      # genstrings outputs UTF-16, so we should try to use this encoding if it turns out to be valid
+      test_message = message.dup
+      return message.encode(Encoding::UTF_8, Encoding::UTF_16) if test_message.force_encoding(Encoding::UTF_16).valid_encoding?
+
+      # replace any invalid with empty string
+      return message.encode(Encoding::UTF_8, invalid: :replace, replace: "")
```
Comment (ainame) [inline fastlane_core/lib/fastlane_core/ui/implementations/shell.rb:161]: Maybe it's worth leaving invalid characters as `�` as people would still want to if there were any chars there?


```suggestion
      return message.encode(Encoding::UTF_8, invalid: :replace)
```

https://ruby-doc.org/core-2.4.0/String.html#method-i-encode

**Review by ainame — COMMENTED** [review]
I could confirm that it's failing on 2.4 on my end but @joshdholtz's fix looks good to me as there would be some histories in Ruby to dig around that but couldn't find it quickly🙂 I have a minor suggestion lastly!

**Review by joshdholtz — APPROVED** [review]
Thanks for fixing this, @sdjmchattie ❤️ Looks like this should be all good to go now. Great work, team!

**Pull request merged**